src/core/sock/sockinfo_ulp.cpp

```diff
@@ -116,10 +116,34 @@ template <typename T> static void dlsym_default(T &ptr, const char *name)
     dlsym_handle(ptr, name, RTLD_DEFAULT);
 }
 
-#define XLIO_TLS_API_FIND(__name) dlsym_default(s_tls_api.__name, #__name);
+static void *openssl_handle = nullptr;
 
-void xlio_tls_api_setup()
+#define XLIO_TLS_API_FIND(__name) dlsym_handle(s_tls_api.__name, #__name, openssl_handle);
+
+inline bool check_openssl_loaded()
+{
+    openssl_handle = dlopen("libssl.so.3", RTLD_NOW | RTLD_GLOBAL);
+    if (!openssl_handle) {
+        vlog_printf(VLOG_DEBUG, "Failed to dlopen libssl.so.3: %s", dlerror());
+        return false;
+    } else {
+        vlog_printf(VLOG_DEBUG, "Successfully loaded libssl.so.3");
+        return true;
+    }
+}
+
+inline void xlio_tls_api_setup()
 {
+    if (openssl_handle) {
+        // OpenSSL symbols is already loaded
+        return;
+    }
+
+    if (!check_openssl_loaded()) {
+        vlog_printf(VLOG_DEBUG, "OpenSSL library not found or failed to load");
+        return;
+    }
+
     XLIO_TLS_API_FIND(EVP_CIPHER_CTX_new);
```

Comment:
I suggest trying the symbols first the old way without openssl_handle as this is a more generic way and does not require loading a specific lib file. And only if it fails then to fallback to dlopen